tesseract-server/src/handlers/logic_layer/aggregate.rs

```diff
@@ -71,124 +133,188 @@ pub fn logic_layer_aggregation(
-                    Ok(time) => time,
-                    Err(err) => return boxed_error(err.to_string())
-                };
+        info!("Query IR: {:?}", query_ir);
 
-                match cube_cache.clone() {
-                    Some(cache) => {
-                        let (cut, cut_value) = match cache.get_time_cut(time) {
-                            Ok(cut) => cut,
-                            Err(err) => return boxed_error(err.to_string())
-                        };
+        let sql = req.state()
+            .backend
+            .generate_sql(query_ir);
 
-                        agg_query.cuts = match agg_query.cuts {
-                            Some(mut cuts) => {
-                                cuts.insert(cut, cut_value);
-                                Some(cuts)
-                            },
-                            None => {
-                                let mut m: HashMap<String, String> = HashMap::new();
-                                m.insert(cut, cut_value);
-                                Some(m)
-                            },
-                        }
-                    },
-                    None => (),
-                };
+        info!("SQL query: {}", sql);
+
+        // Substitute header names (only need to do this once)
+        if final_headers.len() == 0 {
+            for header in &headers {
+                let mut new_header = header.clone();
+
+                for (k, v) in header_map.iter() {
+                    if header.contains(k) {
+                        new_header = new_header.replace(k, v);
+                    }
+                }
+
+                final_headers.push(new_header);
             }
-        },
-        None => (),
-    }
+        }
 
-    info!("Aggregate query: {:?}", agg_query);
+        sql_strings.push(sql);
+    }
 
-    // Turn AggregateQueryOpt into TsQuery
-    let ts_query: Result<TsQuery, _> = agg_query.try_into();
-    let ts_query = match ts_query {
-        Ok(q) => q,
-        Err(err) => return boxed_error(err.to_string())
-    };
+    info!("Headers: {:?}", final_headers);
+
+    // Joins all the futures for each TsQuery
+    let futures: JoinAll<Vec<Box<Future<Item=DataFrame, Error=Error>>>> = join_all(sql_strings
+            .iter()
+            .map(|sql| {
+                req.state()
+                    .backend
+                    .exec_sql(sql.clone())
+            })
+            .collect()
+        );
+
+    // Process data received once all futures are resolved and return response
+    futures
+        .and_then(move |dfs| {
+            let mut final_columns: Vec<Column> = vec![];
+            let num_cols = dfs[0].columns.len();
+
+            for col_i in 0..num_cols {
+                let mut same_type = true;
+                let mut curr_type: i32 = -1;
+
+                for df in &dfs {
+                    let c = &df.columns[col_i];
+                    let i = match c.column_data {
```

Comment:
can you walk me through this section in person? First glance, changing an enum to ints (kind of like C or Go enums) seems strange, but I might not be understanding your needs here.

Also, if it is important, then perhaps some of this can be put into the Dataframe module.